src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -1079,10 +1086,13 @@ trait Contexts { self: Analyzer =>
         defSym = lookupInScope(cx.scope) match {
           case Nil                  => searchPrefix
           case entries @ (hd :: tl) =>
-            // we have a winner: record the symbol depth
-            symbolDepth = (cx.depth - cx.scope.nestingLevel) + hd.depth
-            if (tl.isEmpty) hd.sym
-            else newOverloaded(cx.owner, pre, entries)
+            if (hd.owner eq enclosingPatternCx.scope) NoSymbol
```

Comment:
This comparison looks a bit shady -- could you elaborate? (Why not start outside the pattern context to begin with?) Looks good from the repl, though:

```
Welcome to Scala version 2.11.6-4e70946-SNAPSHOT (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_25).

scala> class A ; val a = new A

scala> (1, a) match { case (a, x : a.type) => (a,x) }
res2: (Int, A) = (1,A@32e1a340)

scala>  (1, a) match { case (a, x : (a.type forSome {type T})) => (a,x) } // type mismatch in 2.11.5
warning: there was one feature warning; re-run with -feature for details
res3: (Int, A) = (1,A@32e1a340)
```